forge/test/mlir/llama/test_llama_inference.py

```diff
@@ -126,3 +127,141 @@ def test_llama_inference_cache_cpu():
+        tt_output = [tt_out.to("cpu") for tt_out in tt_output]
+
+        # Validate TT result with Framework
+        assert all([compare_with_golden_pcc(golden=fw_out, calculated=tt_out, pcc=0.99) for fw_out, tt_out in zip(framework_output, tt_output)])
+        next_token_logits = tt_output[0][:, -1, :]
+        next_token = torch.argmax(next_token_logits, dim=-1)
+
+        if next_token == tokenizer.eos_token_id:
+            break
+
+        input_ids = torch.cat(
+            [input_ids, next_token.unsqueeze(0)], dim=-1
+        )
+
+
+    # Generated text
+    generated_text = tokenizer.decode(input_ids[0], skip_special_tokens=True)
+    print("generated_text=",generated_text)
+
+
+@pytest.mark.xfail()
+def test_llama_inference_cache_tt():
+
+    # Load Llama 3B model and tokenizer
+    model_path = "openlm-research/open_llama_3b"
+    framework_model = load_model(model_path, use_cache=True)
+    tokenizer = LlamaTokenizer.from_pretrained(model_path)
+
+    class Llama_Model_Wrapper(torch.nn.Module):
+        def __init__(self, model):
+            super().__init__()
+            self.model = model
+        def forward(self, inputs):
+            past_key_values_list = inputs[1:]
+            past_key_values = list()
+            for idx in range(0, len(past_key_values_list), 2):
+                past_key_values.append(tuple([past_key_values_list[idx], past_key_values_list[idx + 1]]))
+            model_outputs = self.model(input_ids=inputs[0], past_key_values=tuple(past_key_values))
+            outputs = [model_outputs[0]]
+            for k, v in model_outputs[1]:
+                outputs.append(k)
+                outputs.append(v)
+            return outputs
+
+    llama_model= Llama_Model_Wrapper(framework_model)
+    llama_model.eval()
+
+    # Prepare input sentence
+    prompt = "Q: What is the largest animal?\nA:"
+    inputs = tokenizer(prompt, return_tensors="pt")
+
+    # Run Prefill on CPU with cache to get the initial logits and past key-values
+    logits, past_key_values = framework_model(input_ids=inputs.input_ids, attention_mask=inputs.attention_mask)
+    next_token_logits = logits[:, -1, :]
+    next_token = torch.argmax(next_token_logits, dim=-1)
+
+    generated_tokens = inputs.input_ids
+    generated_tokens = torch.cat(
+        [generated_tokens, next_token.unsqueeze(0)], dim=-1
+    )
+
+    model_inputs = [next_token.unsqueeze(0)]
+    for k, v in past_key_values:
+        model_inputs.append(k)
+        model_inputs.append(v)
+
+
+    # Run decode stage on TT device and generate tokens by passing the last predicted token and the past key values.
+    # untill the a specified maximum number of new tokens is reached or an end-of-sequence token is encountered.
+    max_new_tokens = 32
+    for _ in range(max_new_tokens):
+
+        # CPU Inference
+        framework_output = llama_model(model_inputs)
+
+        # Compile the model
+        compiled_model = forge.compile(llama_model, sample_inputs=[model_inputs])
```

Comment:
Here we compile model in each iteration... @nvukobratTT this will be really slow? Should we fix some max input size ?

Reply:
I agree, we should avoid compiling in the loop (also posted comment for example above).

Reply:
@nvukobratTT @dgolubovicTT 

1. **test_llama_prefill_on_cpu_decode_on_tt_no_cache** - Modified the test to compile outside the loop by making the input with max fixed size (i.e applying padding in the input_ids and additionally passing attention_mask to avoid attention on padding token indicies)

 2. **test_llama_prefill_on_cpu_decode_on_tt_cache** - In the decode, we pass the last predicted token of shape `(batch_size, 1)` **eg: (1, 1)** and past key values of `List[List[torch.Tensor, torch.Tensor]]` and past key values tensor shape will be `(batch_size, num_of_key_values_heads, key_value_seq_len, head_dim)` **eg: (1, 32, 12, 100)** to the LlamaModelWrapper model and the output will be logits of shape (batch_size, 1, vocab_size) **eg: (1, 1, 32000)** and past key values of `List[List[torch.Tensor, torch.Tensor]]` but the returned past key values tensor shape will be `(batch_size, num_of_key_values_heads, key_values_seq_len + 1, head_dim)` **eg: (1, 32, 13, 100)**  Here, you can see in the past key values tensor, key_value_seq_len is changed from 12 to 13 in the [LlamaAttention module](https://github.com/huggingface/transformers/blob/514de24abfd4416aeba6a6455ad5920f57f3567d/src/transformers/models/llama/modeling_llama.py#L374C9-L390C105) by concatenating the current key and value state into the past key values. 
 In this case, how we can compile the model  outside the loop, because in every iteration the key_value_seq_len will changed to key_value_seq_len + 1?

Reply:
Will take a look into the changed PR a bit later, but regarding 2. point, can we also pad KV tensors to the max seq len we defined for inputs?

Reply:
I will look at this PR again later today...

Reply:
I have tried adding padding to the past key values tensor in key_values_seq_len dimension but the generated tokens are different comparing to the past key values tensor without padding.

**Text generated in decode phase:** 

1. Without Past key values padded:
```
Q: What is the largest animal?
A: The blue whale.
Q: What is the largest animal?
A: The blue whale. It is the largest animal on Earth. It is also the largest
```

2. With Past key values padded:
```
Q: What is the largest animal?
A: The largest animal is the blue whale.
Q:
A: What is the largest animal?
A:
A:
A: The blue whale.
```

Reply:
I am looking into how we can solve this. I have read the llama source code [LlamaAttention module](https://github.com/huggingface/transformers/blob/514de24abfd4416aeba6a6455ad5920f57f3567d/src/transformers/models/llama/modeling_llama.py#L374C9-L390C105) and I understand the problem. Will update here when I have some solution... @chandrasekaranpradeep can you provide some branch where you added padding so I can start from there...

Reply:
@chandrasekaranpradeep what values are you using for padding?

Reply:
So padding with 0 should work for calculating attention. However, we have to update kv cache differently than Llama is doing. They are concatenating old K and V states with new. Since out K and V are now constant length = max_seq_len we have to insert new k and v vectors at the end of sequence...

Reply:
> @chandrasekaranpradeep can you provide some branch where you added padding so I can start from there...

Hey @dgolubovicTT, it's already there on PR. Check out this function:
```
test_llama_prefill_on_cpu_decode_on_tt_cache_padding
```

> So padding with 0 should work for calculating attention. However, we have to update kv cache differently than Llama is doing. They are concatenating old K and V states with new. Since out K and V are now constant length = max_seq_len we have to insert new k and v vectors at the end of sequence...

Yea, we'll probably have to manage KV pairs for every iteration in order to populate them in a proper way..

Reply:
Let's see if we can figure out the solution with a current implementation in the next few days. If not, let's wait for other TTNN issues to get resolved, and then push Llama without past cache for a first run. After we finish with ResNet, we should look over PK once again :))